app/src/main/java/ru/spbhse/pocketmagic/Controller.java

```diff
@@ -0,0 +1,536 @@
+        }
+
+        public int getPlayerMP() {
+            return playerMP;
+        }
+
+        public int getMaxHp() {
+            return MAX_HP;
+        }
+
+        public int getMaxMp() {
+            return MAX_MP;
+        }
+
+        synchronized public void initializeCast(String spell) {
+            playerMP -= getCostByName(spell);
+            playerMP = max(0, playerMP);
+        }
+
+        synchronized public void playerSpell(String spell) {
+            playerHP += getHealingByName(spell);
+            playerHP = min(playerHP, MAX_HP);
+            opponentHP -= calcOpponentDamage(spell);
+            opponentHP = max(opponentHP, 0);
+            updateOpponentState(spell);
+        }
+
+        public int calcPlayerDamage(String spell) {
+            int result = getDamageByName(spell);
+            if (playerState == PlayerState.WET && spell.equals("Lightning")) {
+                result = result * SPELL_LENS;
+            }
+            if ((playerState == PlayerState.FREEZING || playerState == PlayerState.FROZEN) && (spell.equals("Breeze"))) {
+                result = SPELL_LENS;
+            }
+            return result;
+        }
+
+        public int calcOpponentDamage(String spell) {
+            int result = getDamageByName(spell);
+            if (opponentState == PlayerState.WET && spell.equals("Lightning")) {
+                result = result * SPELL_LENS;
+            }
+            if ((opponentState == PlayerState.FREEZING || opponentState == PlayerState.FROZEN) && (spell.equals("Breeze"))) {
+                result = SPELL_LENS;
+            }
+            return result;
+        }
+
+        synchronized public void opponentSpell(String spell) {
+            playerHP -= calcPlayerDamage(spell);
+            playerHP = max(playerHP, 0);
+            opponentHP += getHealingByName(spell);
+            opponentHP = min(opponentHP, MAX_HP);
+            updatePlayerState(spell);
+        }
+
+        synchronized public void updatePlayerState(String spell) {
+            if (spell.equals("hide") || spell.equals("after spell")) {
+                playerState = PlayerState.NORMAL;
+            }
+            if (playerState == PlayerState.SUNNY) {
+                return;
+            }
+            if (spell.equals("SunShield")) {
+                playerState = PlayerState.SUNNY;
+            }
+            if (spell.equals("Heal")) {
+                playerState = PlayerState.NORMAL;
+            }
+            if (playerState == PlayerState.NORMAL && spell.equals("Fog")) {
+                playerState = PlayerState.FOG;
+            }
+            if (playerState == PlayerState.FOG && spell.equals("Breeze")) {
+                playerState = PlayerState.WET;
+            }
+            if (playerState == PlayerState.FROZEN && spell.equals("FireBall")) {
+                playerState = PlayerState.FOG;
+            }
+            if (playerState == PlayerState.WET && spell.equals("Freeze")) {
```

Comment:
The problem is that when a player writes something on a screen the GestureLibrary object returns ArrayList with Prediction objects, each of them contains string as a name of recognised gesture. So if we want to use enums instead of strings, we should have a code for matching gesture names and enums. Is it better way?

Reply:
I would create a common routine for the conversion and use it whenever we receive a string from somewhere where it cannot be an enum.